### PR TITLE
Dbz 137 implement log miner emitter

### PR DESCRIPTION
**Pull request by AndreyIg**
Hi, we collected so many changes and we have to sync them somehow.
This code we are currently testing, including performance issues.
Starting from now Danny and myself will work of my fork until you merge the code.
This version does Log Mining without CONTINUOUS_MINE option.
We are mining archived logs on the startup, starting from the offset SCN. After catching up, we are mining online redos only.
We believe this mitigates CONTINUOUS_MINE deprecation.
We did not cover dispatching of DDL and some complicated cases like a DDL during downtime, we will address it later. We did not cover all data types yet.

### CODE REVIEW

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerStreamingChangeEventSource.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+            // 2. Querying LogMinerRowLcr(s) from Log miner while running
+            while (context.isRunning()) {
+                LOGGER.trace("Receiving a change from LogMiner");
+
+                lastProcessedScn = offsetContext.getScn();
+                long currentScn = LogMinerHelper.getCurrentScn(connection);
+                LOGGER.debug("\nstart, lastProcessedScn: " + lastProcessedScn + ", endScn: " + currentScn);
+                if (lastProcessedScn == currentScn) {
+                    LOGGER.debug("No changes to process...pausing for a second");
+                    Metronome metronome = Metronome.sleeper(Duration.ofSeconds(1L), clock);
```
Comment (gunnarmorling) [inline debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerStreamingChangeEventSource.java]: This should use `CommonConnectorConfig.getPollInterval()`. And the metronome should be set up once and stored in a field.

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerStreamingChangeEventSource.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+                debugInfo(connection, "before");
+                res = fetchChangesFromMiningView.executeQuery();
+                debugInfo(connection, "after");
+
+                dispatchChanges(res);
+                offsetContext.setScn(currentScn);
+
+                res.close();
+            }
+        } catch (Throwable e) {
```
Comment (gunnarmorling) [inline debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerStreamingChangeEventSource.java]: Is there any specific need for catching `Throwable`? If not, I'd just let it be passed through.

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerStreamingChangeEventSource.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+            throw new RuntimeException(e);
+        } finally {
+            // 3. disconnect
+            try (Connection connection = jdbcConnection.connection()) {
+                LogMinerHelper.endMining(connection);
+            } catch (SQLException e) {
+                LOGGER.error("Cannot close Log Miner session gracefully: " + e.getMessage());
```
Comment (gunnarmorling) [inline debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerStreamingChangeEventSource.java]: Please keep the original exception:
```suggestion
                LOGGER.error("Cannot close Log Miner session gracefully", e);
```

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerStreamingChangeEventSource.java [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    }
+
+    @Override
+    public void commitOffset(Map<String, ?> offset) {
+        // todo nothing here?
+    }
+
+    private void dispatchChanges(ResultSet res) throws SQLException {
+
+        LOGGER.debug("dispatching changes..");
```
Comment (gunnarmorling) [inline debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerStreamingChangeEventSource.java]: `trace()` probably?

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerStreamingChangeEventSource.java [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+            offsetContext.setTableId(getTableId(logMinerRowLcr));
+            try {
+                LOGGER.debug("Processing DML event {} scn {}", logMinerRowLcr, actualScn);
+                TableId tableId = getTableId(logMinerRowLcr);
+                dispatcher.dispatchDataChangeEvent(
+                        tableId,
+                        new ChangeRecordEmitter(offsetContext, logMinerRowLcr, schema.tableFor(tableId), clock)
+                );
+            } catch (InterruptedException e) {
+                LOGGER.error("Following SCN: " + logMinerRowLcr.getActualScn() + " cannot be dispatched due to the : " + e);
```
Comment (gunnarmorling) [inline debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerStreamingChangeEventSource.java]: `InterruptedException` should rather be let passed up to the caller which handles it appropriately.

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/OracleChangeRecordValueConverter.java [hunk 6]
```diff
@@ -0,0 +1,501 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.connector.oracle.logminer;
```
Comment (gunnarmorling) [inline debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/OracleChangeRecordValueConverter.java]: Why is this under `logminer`. Isn't it generally used?

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/valueholder/LogMinerLcr.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    /**
+     * @return table name
+     */
+    String getObjectName();
+
+    /**
+     * Sets table name
+     * @param name table name
+     */
+    void setObjectName(String name);
```
Comment (gunnarmorling) [inline debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/valueholder/LogMinerLcr.java]: Same question as above about mutability.

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/ChangeRecordEmitter.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+    @Override
+    protected Operation getOperation() {
+        return lcr.getCommandType();
+    }
+
+    @Override
+    protected Object[] getOldColumnValues() {
+        List<LogMinerColumnValue> valueList =  lcr.getOldValues();
+        LogMinerColumnValue[] result = Arrays.copyOf(valueList.toArray(), valueList.size(), LogMinerColumnValue[].class );
```
Comment (gunnarmorling) [inline debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/ChangeRecordEmitter.java]: Can we have `LogMinerRowLcr` return arrays from the get-go, then we'd avoid the copying in the hot code path here.

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/OracleSchemaChangeEventEmitter.java [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+import io.debezium.connector.oracle.BaseOracleSchemaChangeEventEmitter;
+import io.debezium.connector.oracle.OracleOffsetContext;
+import io.debezium.connector.oracle.logminer.valueholder.LogMinerDdlLcr;
+import io.debezium.pipeline.spi.SchemaChangeEventEmitter;
+import io.debezium.relational.TableId;
+
+/**
+ * {@link SchemaChangeEventEmitter} implementation based on Oracle LogMiner utility.
+ */
+public class OracleSchemaChangeEventEmitter extends BaseOracleSchemaChangeEventEmitter {
```
[the remark this answers is not in the document]
Comment (AndreyIg) [reply debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/OracleSchemaChangeEventEmitter.java]: done, will be in the next PR

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/valueholder/LogMinerRowLcr.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+     * the commit scn obtained from a Log Miner entry
+     * @return it's value
+     */
+    long getActualCommitScn();
+
+    /**
+     * sets commit scn obtained from a Log Miner entry
+     * @param actualCommitScn the value
+     */
+    void setActualCommitScn(long actualCommitScn);
```
[the remark this answers is not in the document]
Comment (AndreyIg) [reply debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/valueholder/LogMinerRowLcr.java]: we incorporate SCN info into this object to update offset SCN

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/valueholder/LogMinerRowLcr.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+     * this getter
+     * @return Envelope.Operation enum
+     */
+    Envelope.Operation getCommandType();
+
+    /**
+     * the commit scn obtained from a Log Miner entry
+     * @return it's value
+     */
+    long getActualCommitScn();
```
[the remark this answers is not in the document]
Comment (AndreyIg) [reply debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/valueholder/LogMinerRowLcr.java]: how about BigDecimal?

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/valueholder/LogMinerRowLcrImpl.java [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+    public Timestamp getSourceTime() {
+        return sourceTime;
+    }
+
+    @Override
+    public void setTransactionId(String id) {
+        this.transactionId = id;
+    }
+
+    public void setSourceDatabaseName(String name){
```
[the remark this answers is not in the document]
Comment (AndreyIg) [reply debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/valueholder/LogMinerRowLcrImpl.java]: yes, we can do it, we can save database name in the parser and construct objects with this name.
yes, we are not changing database dynamically

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/xstream/OracleSchemaChangeEventEmitter.java [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+import io.debezium.pipeline.spi.SchemaChangeEventEmitter;
+import io.debezium.relational.TableId;
+import oracle.streams.DDLLCR;
+
+/**
+ * {@link SchemaChangeEventEmitter} implementation based on Oracle.
+ *
+ * @author Gunnar Morling
+ */
+public class OracleSchemaChangeEventEmitter extends BaseOracleSchemaChangeEventEmitter {
```
[the remark this answers is not in the document]
Comment (AndreyIg) [reply debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/xstream/OracleSchemaChangeEventEmitter.java]: we did

File: debezium-connector-oracle/src/test/java/io/debezium/connector/oracle/LogMinerOracleConnectorIT.java [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeFormatterBuilder;
+import java.time.temporal.ChronoField;
+
+/**
+ * This subclasses common OracleConnectorIT for LogMiner adaptor
+ *
+ */
+public class LogMinerOracleConnectorIT extends OracleConnectorIT {
```
[the remark this answers is not in the document]
Comment (AndreyIg) [reply debezium-connector-oracle/src/test/java/io/debezium/connector/oracle/LogMinerOracleConnectorIT.java]: we will review it

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerProcessor.java [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+            boolean tableParsed = false;
+            for  (TableId tableId : schema.getTables().tableIds()) {
+                if (tableId.table().toLowerCase().contains(tableName.toLowerCase())){
+                    tableParsed = true;
+                    break;
+                }
+            }
+
+            if(!tableParsed){
+                LOGGER.debug("table " + tableName + " is not in the whitelist, skipped");
```
[the remark this answers is not in the document]
Comment (AndreyIg) [reply debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerProcessor.java]: sure

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerProcessor.java [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+                continue;
+            }
+
+            //TCL
+            if (operationCode == 7 || operationCode == 36) {
```
[the remark this answers is not in the document]
Comment (AndreyIg) [reply debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerProcessor.java]: we were planning to do this plus all other constants

**Comment by AndreyIg** [comment]
we are currently addressing one serious problem: some changes are missed due to the REDO LOG flush timing.
When we address it, we will return to all those above

**Comment by AndreyIg** [comment]
We got in sync with Dany again. In this commit we addressed 
- a few community comments,
- implemented sdo_geometry, CLOB and BLOB simple conversion to string,
- make it working with no container database
- fixed timestamp problems,
- handled redo log files switches, which require rebuilding dictionary and readding files
- parsing table alias
- replaced metadata usage with sql queries for couple of cases, still have to address columns metadata
- handled UNUSED state of redo log files

**Comment by Daniel-Yosief** [comment]
Added TransactionalBuffer concept from @jchipmunk.

**Comment by Daniel-Yosief** [comment]
Fix logic to update SCN on offset context

**Comment by gunnarmorling** [comment]
Hey @Daniel-Yosief and @AndreyIg, I'm back from three weeks of vacation :) How are things for you? Are you planning to more work to this right now? Or should I do another review pass? Thanks!

**Comment by gunnarmorling** [comment]
Btw. could you rebase this PR to the latest master? It has some conflicts.

Also, this blog post could be interesting to read: https://medium.com/@FranckPachot/ideas-for-event-sourcing-in-oracle-d4e016e90af6. It discusses some CDC options for Oracle, but seems to arrive at the same conclusion as you do: use LogMiner without "continuous mine".

**Comment by gunnarmorling** [comment]
@AndreyIg, @Daniel-Yosief, howdy? Should we sync on this PR again?

**Comment by kpmeen** [comment]
Has there been any (hidden) progress on this PR? As far as I can tell nothing has happened in the last 4 months or so.

**Comment by gunnarmorling** [comment]
@kpmeen, yes, @AndreyIg is working on this; I'll refer to him for the details, but this PR should be updated within the next few weeks.

**Comment by kpmeen** [comment]
That's great new @gunnarmorling . Do you have any (rough) estimates for when this might be merged to master and made available in a release (or RC)? If you need, I'd love to help testing it to speed things up.

**Comment by jchipmunk** [comment]
@Daniel-Yosief, @AndreyIg, I read your MR and I have the following questions:
1. Why aren’t you mine data from archive logs?
2. How do you resolve the problem of cyclic overwriting of online log files?

>LogMiner can mine online redo logs. However, if the CONTINUOUS_MINE option is not specified, it is possible that the database is writing to the online redo log file at the same time that LogMiner is reading the online redo log file. If a log switch occurs while LogMiner is reading an online redo log file, the database will overwrite what LogMiner is attempting to read. The data that LogMiner returns if the file it is trying to read gets overwritten by the database is unpredictable.

For more information, see [DBMS_LOGMNR](https://docs.oracle.com/database/121/ARPLS/d_logmnr.htm#ARPLS66809).

**Comment by AndreyIg** [comment]
Hi, Andrey
1. We are currently OK to mine up to 1 hour back in history. We assume the
connector will not be down for a longer time. We require 7 redo groups 0.5G
each file.
For our load it is sufficient (15min switch * 7)
Since continious_mine is deprecated (and throws on Oracle 19C which is
supposed to be the longest supported version) we add/remove files manually.
For the first release we want to have a simple algorithm as much as
possible. No archives, no DDL
2. When next switch has happen, it turns state to "active" to archive it's
content. It does not get overwritten until end of full cycle (7 switches in
this case). The file changes its states: current --> active -> inactive.
The mining gets paused when file is in active state (a few seconds).

We are in testing phase and already passed many scenarios. Currently we are
addressing PGA leak on windows installation.
I included Andrew - the third Andrey :)  who is our DBA. Please correct us
if we are wrong.

Thank you,

**Comment by kpmeen** [comment]
Hello @AndreyIg, thanks for your effort on this! I'm looking forward to try this connector. Do you have any rough estimates on how far away you are from being able to merge and release? Will there be a beta or pre-release that I could try out?

**Comment by AndreyIg** [comment]
Hi, this version is in pre-production testing phase.
I would appreciate if you can try and test it too.
Currently I'm waiting Gunnar Morling's review, suggestions and help.
I have to capacity to merge it soon.
Thank you,
Andrey

**Comment by thake** [comment]
I just now noticed this pull request. In the last weeks I've also written a Kafka Connector with logminer support (https://github.com/thake/logminer-kafka-connect). As your attempt seems to be more mature, I would be happy to contribute.

I noticed that you are just reading committed records. How do you cope with very large transactions? Logminer might just give up due to memory constraints. Also queries against V$LOGMNR_CONTENTS might take significantly longer.

> If the COMMITTED_DATA_ONLY option is specified and you issue a query, then LogMiner stages all redo records within a single transaction in memory until LogMiner finds the commit record for that transaction. Therefore, it is possible to exhaust memory, in which case an "Out of Memory" error will be returned. If this occurs, then you must restart LogMiner without the COMMITTED_DATA_ONLY option specified and reissue the query. (https://docs.oracle.com/database/121/SUTIL/GUID-6A2398F7-D484-495A-8AD2-0A6B34C03536.htm#SUTIL1574)

Because of this I switched to reading uncommitted data. Uncommitted records will be stored to a memory mapped file and will be only processed if a commit occurs. Have a look at https://github.com/thake/logminer-kafka-connect/blob/master/src/main/kotlin/com/github/thake/logminer/kafka/connect/logminer/Transaction.kt.

**Comment by FrankMormino** [comment]
@thake @AndreyIg - awesome work here -  thanks for your effort on this! I'm looking forward to try this connector. Do you have any rough estimates on how far away you are from being able to merge and release? Will there be a beta or pre-release that I could try out?

**Comment by AndreyIg** [comment]
Sorry for the late reply.
I'm really busy.
The connector is passing last stages of scalability and resiliency testing.
I'll update my for as soon as possible. I hope next week.
Thanks.

**Comment by FrankMormino** [comment]
@thake @AndreyIg - do you have any notes for anyone who wants to test this out for you? I have a scenario where we could utilize this and I'm trying to get this to work against oracle logminer instance 11.2.0.3 - I have spent day or two navigating the documentation from debezium to debezium-incubator and want to make sure I am configuring and building this connector correctly. Is there anywhere you can point me for this or GOTCHAs you've ran into while you tested this out?

**Comment by crognon** [comment]
Hi,

Thank you for the good job ! how can we test your work to report feedback here ?

thanks
Cyril

**Comment by AndreyIg** [comment]
Hi, We are planning to merge it to the Debezuim base this or maybe next
week.

**Comment by Naros** [comment]
Superseded by PR #185.

**Pull request closed without merging**